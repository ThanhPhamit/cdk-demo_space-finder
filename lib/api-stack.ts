import * as cdk from 'aws-cdk-lib';
import * as apigateway from 'aws-cdk-lib/aws-apigateway';
import { IUserPool } from 'aws-cdk-lib/aws-cognito';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';

interface ApiStackProps extends cdk.StackProps {
  spacesLambda: lambda.Function;
  userPool: IUserPool;
}

export class ApiStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: ApiStackProps) {
    super(scope, id, props);

    // Create API Gateway
    const api = new apigateway.RestApi(this, 'SpaceApi', {
      restApiName: 'Space Service',
      description: 'API Gateway for Space Lambda function',
      defaultCorsPreflightOptions: {
        allowOrigins: apigateway.Cors.ALL_ORIGINS,
        allowMethods: apigateway.Cors.ALL_METHODS,
      },
    });

    // Add Cognito User Pool Authorizer
    const authorizer = new apigateway.CognitoUserPoolsAuthorizer(
      this,
      'SpaceUserPoolAuthorizer',
      {
        cognitoUserPools: [props.userPool],
        identitySource: 'method.request.header.Authorization',
      },
    );

    authorizer._attachToApi(api);

    const optionsWithAuth: apigateway.MethodOptions = {
      authorizationType: apigateway.AuthorizationType.COGNITO,
      authorizer: { authorizerId: authorizer.authorizerId },
    };

    // Create Lambda integrations
    const spacesIntegration = new apigateway.LambdaIntegration(
      props.spacesLambda,
      {
        requestTemplates: { 'application/json': '{ "statusCode": "200" }' },
      },
    );

    // Add resources and methods
    const spacesResource = api.root.addResource('spaces');
    spacesResource.addMethod('GET', spacesIntegration, optionsWithAuth);
    spacesResource.addMethod('POST', spacesIntegration, optionsWithAuth);
    spacesResource.addMethod('PUT', spacesIntegration, optionsWithAuth);
    spacesResource.addMethod('DELETE', spacesIntegration, optionsWithAuth);

    // Output the API Gateway URL
    new cdk.CfnOutput(this, 'ApiGatewayUrl', {
      value: api.url,
      description: 'URL of the API Gateway',
    });

    // Output specific endpoints
    new cdk.CfnOutput(this, 'SpacesEndpoint', {
      value: `${api.url}spaces`,
      description: 'Spaces endpoint URL',
    });
  }
}
