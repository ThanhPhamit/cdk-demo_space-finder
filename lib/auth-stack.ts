import * as cdk from 'aws-cdk-lib';
import {
  CfnIdentityPool,
  CfnIdentityPoolRoleAttachment,
  CfnUserPoolGroup,
  UserPool,
  UserPoolClient,
} from 'aws-cdk-lib/aws-cognito';
import {
  Effect,
  FederatedPrincipal,
  PolicyStatement,
  Role,
} from 'aws-cdk-lib/aws-iam';
import { IBucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

interface AuthStackProps extends cdk.StackProps {
  photosBucket: IBucket;
}

export class AuthStack extends cdk.Stack {
  public userPool: UserPool;
  private userPoolClient: UserPoolClient;
  private identityPool: CfnIdentityPool;
  private authenticatedRole: Role;
  private unauthenticatedRole: Role;
  private adminRole: Role;

  constructor(scope: Construct, id: string, props: AuthStackProps) {
    super(scope, id, props);

    this.createUserPool();
    this.createUserPoolClient();
    this.createIdentityPool();
    this.createRoles(props.photosBucket);
    this.attachRolesToIdentityPool();
    this.createAdminGroup();
  }

  private createUserPool() {
    this.userPool = new UserPool(this, 'SpaceUserPool', {
      selfSignUpEnabled: true,
      signInAliases: { email: true, username: true },
    });

    new cdk.CfnOutput(this, 'SpaceUserPoolId', {
      value: this.userPool.userPoolId,
    });
  }

  private createUserPoolClient() {
    this.userPoolClient = new UserPoolClient(this, 'SpaceUserPoolClient', {
      userPool: this.userPool,
      authFlows: {
        adminUserPassword: true,
        custom: true,
        userPassword: true,
        userSrp: true,
      },
    });

    new cdk.CfnOutput(this, 'SpaceUserPoolClientId', {
      value: this.userPoolClient.userPoolClientId,
    });
  }

  private createAdminGroup() {
    new CfnUserPoolGroup(this, 'SpaceAdminGroup', {
      groupName: 'admins',
      userPoolId: this.userPool.userPoolId,
      roleArn: this.adminRole?.roleArn,
    });
  }

  private createIdentityPool() {
    this.identityPool = new CfnIdentityPool(this, 'SpaceIdentityPool', {
      allowUnauthenticatedIdentities: false,
      cognitoIdentityProviders: [
        {
          clientId: this.userPoolClient.userPoolClientId,
          providerName: this.userPool.userPoolProviderName,
        },
      ],
    });
    new cdk.CfnOutput(this, 'SpaceIdentityPoolId', {
      value: this.identityPool.ref,
    });
  }

  private createRoles(photosBucket: IBucket) {
    this.authenticatedRole = new Role(this, 'SpaceCognitoAuthRole', {
      assumedBy: new FederatedPrincipal(
        'cognito-identity.amazonaws.com',
        {
          StringEquals: {
            'cognito-identity.amazonaws.com:aud': this.identityPool.ref,
          },
          'ForAnyValue:StringLike': {
            'cognito-identity.amazonaws.com:amr': 'authenticated',
          },
        },
        'sts:AssumeRoleWithWebIdentity',
      ),
    });

    this.unauthenticatedRole = new Role(this, 'SpaceCognitoUnauthRole', {
      assumedBy: new FederatedPrincipal(
        'cognito-identity.amazonaws.com',
        {
          StringEquals: {
            'cognito-identity.amazonaws.com:aud': this.identityPool.ref,
          },
          'ForAnyValue:StringLike': {
            'cognito-identity.amazonaws.com:amr': 'unauthenticated',
          },
        },
        'sts:AssumeRoleWithWebIdentity',
      ),
    });

    this.adminRole = new Role(this, 'SpaceCognitoAdminRole', {
      assumedBy: new FederatedPrincipal(
        'cognito-identity.amazonaws.com',
        {
          StringEquals: {
            'cognito-identity.amazonaws.com:aud': this.identityPool.ref,
          },
          'ForAnyValue:StringLike': {
            'cognito-identity.amazonaws.com:amr': 'authenticated',
          },
        },
        'sts:AssumeRoleWithWebIdentity',
      ),
    });

    this.adminRole.addToPolicy(
      new PolicyStatement({
        effect: Effect.ALLOW,
        actions: ['s3:ListAllMyBuckets'],
        resources: ['*'],
      }),
    );

    this.adminRole.addToPolicy(
      new PolicyStatement({
        effect: Effect.ALLOW,
        actions: [
          's3:PutObject',
          's3:PutObjectAcl',
          's3:GetObject',
          's3:DeleteObject',
        ],
        resources: [photosBucket.arnForObjects('*')],
      }),
    );

    this.adminRole.addToPolicy(
      new PolicyStatement({
        effect: Effect.ALLOW,
        actions: ['s3:ListBucket'],
        resources: [photosBucket.bucketArn],
      }),
    );
  }

  private attachRolesToIdentityPool() {
    new CfnIdentityPoolRoleAttachment(this, 'SpaceIdentityPoolRoleAttachment', {
      identityPoolId: this.identityPool.ref,
      roles: {
        authenticated: this.authenticatedRole.roleArn,
        unauthenticated: this.unauthenticatedRole.roleArn,
      },
      roleMappings: {
        adminsMapping: {
          type: 'Token',
          ambiguousRoleResolution: 'AuthenticatedRole',
          identityProvider: `${this.userPool.userPoolProviderName}:${this.userPoolClient.userPoolClientId}`,
          rulesConfiguration: {
            rules: [
              {
                claim: 'cognito:groups',
                matchType: 'Contains',
                value: 'admins',
                roleArn: this.adminRole.roleArn,
              },
            ],
          },
        },
      },
    });
  }
}
