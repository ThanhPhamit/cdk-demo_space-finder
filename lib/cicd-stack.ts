import * as cdk from 'aws-cdk-lib';
import * as apigateway from 'aws-cdk-lib/aws-apigateway';
import { IUserPool } from 'aws-cdk-lib/aws-cognito';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import {
  CodePipeline,
  CodePipelineSource,
  ShellStep,
} from 'aws-cdk-lib/pipelines';
import { Construct } from 'constructs';

interface CicdStackProps extends cdk.StackProps {}

export class CicdStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: CicdStackProps) {
    super(scope, id, props);

    new CodePipeline(this, 'Pipeline', {
      pipelineName: 'CICDPipeline',
      synth: new ShellStep('Synth', {
        input: CodePipelineSource.gitHub(
          'ThanhPhamit/cdk-demo_space-finder',
          'chore/cicd/cicd-pratice',
        ),
        commands: [
          // use yarn to install reproducibly, build, and run cdk synth via yarn
          'yarn install --frozen-lockfile',
          'yarn tsc --skipLibCheck',  // Skip lib check to avoid @smithy/core type errors
          'yarn cdk synth',
        ],
        primaryOutputDirectory: 'cdk.out',
      }),
    });
  }
}
