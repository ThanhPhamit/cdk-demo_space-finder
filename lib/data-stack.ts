import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import { AttributeType, ITable, Table } from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';
import { getSuffixFromStack } from '../util';
import {
  Bucket,
  HttpMethods,
  IBucket,
  ObjectOwnership,
} from 'aws-cdk-lib/aws-s3';

interface DataStackProps extends StackProps {
  cloudfrontDomain: string; // Made required, not optional
}

export class DataStack extends Stack {
  public readonly spacesTable: ITable;
  public readonly deploymentBucket: IBucket;
  public readonly photosBucket: IBucket;

  constructor(scope: Construct, id: string, props: DataStackProps) {
    super(scope, id, props);

    const suffix = getSuffixFromStack(this);

    this.spacesTable = new Table(this, 'SpacesTable', {
      partitionKey: { name: 'id', type: AttributeType.STRING },
      tableName: 'SpacesTable-' + suffix,
    });

    this.deploymentBucket = new Bucket(this, 'SpacesFinderUiDeploymentBucket', {
      bucketName: `space-finder-ui-deployment-bucket-${suffix}`,
    });

    // Validate that CloudFront domain is provided
    if (!props.cloudfrontDomain) {
      throw new Error(
        'CloudFront domain is required for CORS configuration. Cannot allow all origins (*) for security reasons.',
      );
    }

    // TODO: Remove localhost origin in production
    // Use the specific CloudFront domain for CORS
    const allowedOrigins = [
      `https://${props.cloudfrontDomain}`,
      // 'http://localhost:5173',
    ];

    this.photosBucket = new Bucket(this, 'SpacesPhotosBucket', {
      bucketName: `space-finder-photos-bucket-${suffix}`,
      cors: [
        {
          allowedMethods: [HttpMethods.HEAD, HttpMethods.GET, HttpMethods.PUT],
          allowedOrigins: allowedOrigins,
          allowedHeaders: ['*'],
        },
      ],
      blockPublicAccess: {
        blockPublicAcls: false,
        blockPublicPolicy: false,
        ignorePublicAcls: false,
        restrictPublicBuckets: false,
      },
      objectOwnership: ObjectOwnership.OBJECT_WRITER,
    });

    new CfnOutput(this, 'SpaceFinderPhotosBucketName', {
      value: this.photosBucket.bucketName,
      description: 'The name of the S3 bucket for storing space photos',
    });
  }
}
