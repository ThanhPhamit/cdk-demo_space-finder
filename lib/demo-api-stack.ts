import * as cdk from 'aws-cdk-lib';
import * as apigateway from 'aws-cdk-lib/aws-apigateway';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';

interface DemoApiStackProps extends cdk.StackProps {
  helloLambda: lambda.Function;
  pythonLambda: lambda.Function;
}

export class DemoApiStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: DemoApiStackProps) {
    super(scope, id, props);

    // Create API Gateway
    const api = new apigateway.RestApi(this, 'SpaceApi', {
      restApiName: 'Demo Space Service',
      description: 'API Gateway for Demo Space Lambda function',
      defaultCorsPreflightOptions: {
        allowOrigins: apigateway.Cors.ALL_ORIGINS,
        allowMethods: apigateway.Cors.ALL_METHODS,
      },
    });

    // Create Lambda integrations
    const helloIntegration = new apigateway.LambdaIntegration(
      props.helloLambda,
      {
        requestTemplates: { 'application/json': '{ "statusCode": "200" }' },
      },
    );

    const pythonIntegration = new apigateway.LambdaIntegration(
      props.pythonLambda,
      {
        requestTemplates: { 'application/json': '{ "statusCode": "200" }' },
      },
    );

    // Add resources and methods
    const helloResource = api.root.addResource('hello');
    helloResource.addMethod('GET', helloIntegration);

    const helloPythonResource = api.root.addResource('hello-python');
    helloPythonResource.addMethod('GET', pythonIntegration);

    // Add a catch-all proxy resource for additional paths
    const proxyResource = api.root.addResource('{proxy+}');
    proxyResource.addMethod('ANY', helloIntegration);

    // Output the API Gateway URL
    new cdk.CfnOutput(this, 'ApiGatewayUrl', {
      value: api.url,
      description: 'URL of the API Gateway',
    });

    // Output specific endpoints
    new cdk.CfnOutput(this, 'HelloEndpoint', {
      value: `${api.url}hello`,
      description: 'Hello TypeScript endpoint URL',
    });

    new cdk.CfnOutput(this, 'HelloPythonEndpoint', {
      value: `${api.url}hello-python`,
      description: 'Hello Python endpoint URL',
    });
  }
}
