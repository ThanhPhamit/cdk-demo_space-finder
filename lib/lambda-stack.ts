import * as cdk from 'aws-cdk-lib';
import { ITable } from 'aws-cdk-lib/aws-dynamodb';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { PythonFunction } from '@aws-cdk/aws-lambda-python-alpha';
import { Construct } from 'constructs';
import { join } from 'path';
import { Effect, PolicyStatement } from 'aws-cdk-lib/aws-iam';

interface LambdaStackProps extends cdk.StackProps {
  spacesTable: ITable;
}

export class LambdaStack extends cdk.Stack {
  public readonly helloLambda: lambda.Function;
  public readonly pythonLambda: lambda.Function;

  constructor(scope: Construct, id: string, props?: LambdaStackProps) {
    super(scope, id, props);

    this.helloLambda = new NodejsFunction(this, 'HelloLambda', {
      runtime: lambda.Runtime.NODEJS_20_X,
      handler: 'handler',
      entry: join(__dirname, '..', 'lambda', 'nodejs', 'hello.ts'),
      description: 'A simple Lambda function that logs events',
      timeout: cdk.Duration.minutes(5),
      environment: {
        SPACES_TABLE_NAME: props?.spacesTable.tableName || '',
      },
    });

    this.helloLambda.addToRolePolicy(
      new PolicyStatement({
        effect: Effect.ALLOW,
        actions: ['s3:ListAllMyBuckets', 's3:ListBucket'],
        resources: ['arn:aws:s3:::*'], // bad practice, should be more specific
      }),
    );

    this.pythonLambda = new PythonFunction(this, 'PythonHelloLambda', {
      runtime: lambda.Runtime.PYTHON_3_12,
      entry: join(__dirname, '..', 'lambda', 'python'),
      handler: 'handler',
      description: 'A simple Python Lambda function',
      timeout: cdk.Duration.minutes(5),
      environment: {
        SPACES_TABLE_NAME: props?.spacesTable.tableName || '',
      },
    });

    // Output Lambda function ARNs
    new cdk.CfnOutput(this, 'HelloLambdaFunctionArn', {
      value: this.helloLambda.functionArn,
      description: 'ARN of the Hello Lambda function',
    });

    new cdk.CfnOutput(this, 'PythonLambdaFunctionArn', {
      value: this.pythonLambda.functionArn,
      description: 'ARN of the Python Hello Lambda function',
    });

    // Output CloudWatch Log Group Names
    new cdk.CfnOutput(this, 'HelloLambdaLogGroup', {
      value: this.helloLambda.logGroup.logGroupName,
      description: 'CloudWatch Log Group for Hello Lambda',
    });

    new cdk.CfnOutput(this, 'PythonLambdaLogGroup', {
      value: this.pythonLambda.logGroup.logGroupName,
      description: 'CloudWatch Log Group for Python Lambda',
    });

    // Output CloudWatch Log Group ARNs
    new cdk.CfnOutput(this, 'HelloLambdaLogGroupArn', {
      value: this.helloLambda.logGroup.logGroupArn,
      description: 'CloudWatch Log Group ARN for Hello Lambda',
    });

    new cdk.CfnOutput(this, 'PythonLambdaLogGroupArn', {
      value: this.pythonLambda.logGroup.logGroupArn,
      description: 'CloudWatch Log Group ARN for Python Lambda',
    });
  }
}
