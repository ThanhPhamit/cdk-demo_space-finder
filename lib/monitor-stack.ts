import * as cdk from 'aws-cdk-lib';
import { SnsAction } from 'aws-cdk-lib/aws-cloudwatch-actions';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Topic } from 'aws-cdk-lib/aws-sns';
import { LambdaSubscription } from 'aws-cdk-lib/aws-sns-subscriptions';
import { Construct } from 'constructs';
import { join } from 'path';

interface MonitorStackProps extends cdk.StackProps {}

export class MonitorStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: MonitorStackProps) {
    super(scope, id, props);

    const slackWebhookUrl = process.env.SLACK_WEBHOOK_URL;
    if (!slackWebhookUrl) {
      throw new Error(
        'SLACK_WEBHOOK_URL environment variable is required for MonitorStack',
      );
    }

    const webHookLambda = new NodejsFunction(this, 'WebHookLambda', {
      runtime: cdk.aws_lambda.Runtime.NODEJS_20_X,
      handler: 'handler',
      entry: join(__dirname, '..', 'services', 'monitor', 'handler.ts'),
      environment: {
        SLACK_WEBHOOK_URL: slackWebhookUrl,
      },
    });

    const alarmTopic = new Topic(this, 'AlarmTopic', {
      displayName: 'AlarmTopic',
      topicName: 'AlarmTopic',
    });

    alarmTopic.addSubscription(new LambdaSubscription(webHookLambda));

    const spacesApi4xxAlarm = new cdk.aws_cloudwatch.Alarm(
      this,
      'SpacesApi4xxAlarm',
      {
        metric: new cdk.aws_cloudwatch.Metric({
          namespace: 'AWS/ApiGateway',
          metricName: '4XXError',
          dimensionsMap: {
            ApiName: 'Space Service',
          },
          statistic: 'Sum',
          period: cdk.Duration.minutes(5),
        }),
        threshold: 1,
        evaluationPeriods: 1,
        alarmDescription: 'Alarm for 4XX errors in Space Service API Gateway',
        alarmName: 'SpacesApi4xxAlarm',
      },
    );

    const spacesApi5xxAlarm = new cdk.aws_cloudwatch.Alarm(
      this,
      'SpacesApi5xxAlarm',
      {
        metric: new cdk.aws_cloudwatch.Metric({
          namespace: 'AWS/ApiGateway',
          metricName: '5XXError',
          dimensionsMap: {
            ApiName: 'Space Service',
          },
          statistic: 'Sum',
          period: cdk.Duration.minutes(5),
        }),
        threshold: 1,
        evaluationPeriods: 1,
        alarmDescription: 'Alarm for 5XX errors in Space Service API Gateway',
        alarmName: 'SpacesApi5xxAlarm',
      },
    );

    const topicAction = new SnsAction(alarmTopic);

    spacesApi4xxAlarm.addAlarmAction(topicAction);
    spacesApi4xxAlarm.addOkAction(topicAction);
    spacesApi5xxAlarm.addAlarmAction(topicAction);
  }
}
