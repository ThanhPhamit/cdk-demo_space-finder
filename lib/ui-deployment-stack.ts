import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import { Bucket, IBucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';
import { getSuffixFromStack } from '../util';
import { join } from 'path';
import { existsSync } from 'fs';
import { BucketDeployment, Source } from 'aws-cdk-lib/aws-s3-deployment';
import { Distribution } from 'aws-cdk-lib/aws-cloudfront';
import { S3Origin } from 'aws-cdk-lib/aws-cloudfront-origins';

interface UiDeploymentStackProps extends StackProps {
  deploymentBucket: IBucket;
}

export class UiDeploymentStack extends Stack {
  public readonly distribution: Distribution;

  constructor(scope: Construct, id: string, props: UiDeploymentStackProps) {
    super(scope, id, props);

    const uiDir = join(__dirname, '..', '..', 'space-finder-frontend', 'dist');
    if (!existsSync(uiDir)) {
      throw new Error(
        `UI directory does not exist: ${uiDir}. Please build the frontend first.`,
      );
    }

    new BucketDeployment(this, 'SpacesFinderUiDeployment', {
      destinationBucket: props.deploymentBucket,
      sources: [Source.asset(uiDir)],
    });

    const s3Origin = new S3Origin(props.deploymentBucket);

    this.distribution = new Distribution(this, 'SpacesFinderDistribution', {
      defaultRootObject: 'index.html',
      defaultBehavior: {
        origin: s3Origin,
      },
    });

    new CfnOutput(this, 'DistributionDomainName', {
      value: this.distribution.domainName,
      description: 'The domain name of the CloudFront distribution',
    });
  }
}
