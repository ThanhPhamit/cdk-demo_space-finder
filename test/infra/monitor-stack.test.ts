import { App } from 'aws-cdk-lib';
import { MonitorStack } from '../../lib/monitor-stack';
import { Capture, Match, Template } from 'aws-cdk-lib/assertions';
import 'dotenv/config';

const env = {
  account: process.env.AWS_ACCOUNT_ID,
  region: process.env.AWS_REGION,
};

describe('Initial Test', () => {
  let monitorStackTemplate: Template;
  beforeAll(() => {
    if (!env.account || !env.region) {
      throw new Error(
        'AWS_ACCOUNT_ID and AWS_REGION environment variables must be set for tests',
      );
    }

    const testApp = new App({
      outdir: 'cdk.out',
    });

    const monitorStack = new MonitorStack(testApp, 'MonitorStack', { env });
    monitorStackTemplate = Template.fromStack(monitorStack);
  });

  test('Lambda function properties', () => {
    monitorStackTemplate.hasResourceProperties('AWS::Lambda::Function', {
      Handler: 'index.handler',
      Runtime: 'nodejs20.x',
    });
  });

  test('SNS topic properties', () => {
    monitorStackTemplate.hasResourceProperties('AWS::SNS::Topic', {
      DisplayName: 'AlarmTopic',
      TopicName: 'AlarmTopic',
    });
  });

  // CDK matcher
  test('SNS subscription properties - with matchers', () => {
    monitorStackTemplate.hasResourceProperties(
      'AWS::SNS::Subscription',
      Match.objectEquals({
        Protocol: 'lambda',
        TopicArn: {
          Ref: Match.stringLikeRegexp('AlarmTopic'),
        },
        Endpoint: {
          'Fn::GetAtt': [Match.stringLikeRegexp('WebHookLambda'), 'Arn'],
        },
      }),
    );
  });

  test('SNS subscription properties - with exact values', () => {
    const snsTopic = monitorStackTemplate.findResources('AWS::SNS::Topic');
    const snsTopicName = Object.keys(snsTopic)[0];

    const lambdaFunction = monitorStackTemplate.findResources(
      'AWS::Lambda::Function',
    );
    const lambdaFunctionName = Object.keys(lambdaFunction)[0];

    monitorStackTemplate.hasResourceProperties(
      'AWS::SNS::Subscription',
      Match.objectEquals({
        Protocol: 'lambda',
        TopicArn: {
          Ref: snsTopicName,
        },
        Endpoint: {
          'Fn::GetAtt': [lambdaFunctionName, 'Arn'],
        },
      }),
    );
  });

  // Using SDK Captures
  test('CloudWatch alarm actions', () => {
    const alarmActionsCapture = new Capture();
    monitorStackTemplate.hasResourceProperties('AWS::CloudWatch::Alarm', {
      AlarmActions: alarmActionsCapture,
    });

    expect(alarmActionsCapture.asArray()).toEqual([
      {
        Ref: expect.stringMatching(/^AlarmTopic/),
      },
    ]);
  });

  // Snapshot testing
  test('Monitor stack snapshot', () => {
    expect(monitorStackTemplate.toJSON()).toMatchSnapshot();
  });

  test('Lambda stack snapshot', () => {
    const lambda = monitorStackTemplate.findResources('AWS::Lambda::Function');
    expect(lambda).toMatchSnapshot();
  });

  test('SNS Topic stack snapshot', () => {
    const snsTopic = monitorStackTemplate.findResources('AWS::SNS::Topic');
    expect(snsTopic).toMatchSnapshot();
  });
});
