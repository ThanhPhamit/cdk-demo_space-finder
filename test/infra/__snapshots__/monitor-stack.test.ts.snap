// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Initial Test Lambda stack snapshot 1`] = `
{
  "WebHookLambdaC2AD15FD": {
    "DependsOn": [
      "WebHookLambdaServiceRole1D0B8EE0",
    ],
    "Properties": {
      "Code": {
        "S3Bucket": "cdk-hnb659fds-assets-637423473957-ap-southeast-1",
        "S3Key": "aabcfa999d0a10dfd492a499b1db74c1fb0885d991fc1b7b766f4e0fd5101bd5.zip",
      },
      "Environment": {
        "Variables": {
          "SLACK_WEBHOOK_URL": "PLACEHOLDER_SLACK_WEBHOOK_URL",
        },
      },
      "Handler": "index.handler",
      "Role": {
        "Fn::GetAtt": [
          "WebHookLambdaServiceRole1D0B8EE0",
          "Arn",
        ],
      },
      "Runtime": "nodejs20.x",
    },
    "Type": "AWS::Lambda::Function",
  },
}
`;

exports[`Initial Test Monitor stack snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AlarmTopicD01E77F9": {
      "Properties": {
        "DisplayName": "AlarmTopic",
        "TopicName": "AlarmTopic",
      },
      "Type": "AWS::SNS::Topic",
    },
    "SpacesApi4xxAlarm929EC379": {
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "AlarmTopicD01E77F9",
          },
        ],
        "AlarmDescription": "Alarm for 4XX errors in Space Service API Gateway",
        "AlarmName": "SpacesApi4xxAlarm",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "ApiName",
            "Value": "Space Service",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "4XXError",
        "Namespace": "AWS/ApiGateway",
        "OKActions": [
          {
            "Ref": "AlarmTopicD01E77F9",
          },
        ],
        "Period": 300,
        "Statistic": "Sum",
        "Threshold": 1,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "SpacesApi5xxAlarm01366A76": {
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "AlarmTopicD01E77F9",
          },
        ],
        "AlarmDescription": "Alarm for 5XX errors in Space Service API Gateway",
        "AlarmName": "SpacesApi5xxAlarm",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "ApiName",
            "Value": "Space Service",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "5XXError",
        "Namespace": "AWS/ApiGateway",
        "Period": 300,
        "Statistic": "Sum",
        "Threshold": 1,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "WebHookLambdaAlarmTopicD6696FB1": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "WebHookLambdaC2AD15FD",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": {
          "Ref": "AlarmTopicD01E77F9",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "WebHookLambdaAllowInvokeMonitorStackAlarmTopic9F999C82C57D735E": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "WebHookLambdaC2AD15FD",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Ref": "AlarmTopicD01E77F9",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "WebHookLambdaC2AD15FD": {
      "DependsOn": [
        "WebHookLambdaServiceRole1D0B8EE0",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-637423473957-ap-southeast-1",
          "S3Key": "aabcfa999d0a10dfd492a499b1db74c1fb0885d991fc1b7b766f4e0fd5101bd5.zip",
        },
        "Environment": {
          "Variables": {
            "SLACK_WEBHOOK_URL": "PLACEHOLDER_SLACK_WEBHOOK_URL",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "WebHookLambdaServiceRole1D0B8EE0",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "WebHookLambdaServiceRole1D0B8EE0": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`Initial Test SNS Topic stack snapshot 1`] = `
{
  "AlarmTopicD01E77F9": {
    "Properties": {
      "DisplayName": "AlarmTopic",
      "TopicName": "AlarmTopic",
    },
    "Type": "AWS::SNS::Topic",
  },
}
`;
